### ISSUE TITLE
[Harvesting - Scheduling] Implement a configurable dynamic scheduler

### ISSUE THREAD

**Issue by giohappy** [geosolutions-it/nexus-geonode#187]

**Comment by ricardogsilva** [comment]
I have already implemented a way to have dynamic scheduling with celery beat and django. The implementation is present in geonode/geonode#7317

It is using a similar pattern as the one described in this article:

https://medium.com/swlh/dynamic-task-scheduling-with-django-celery-beat-f2591d52e15

A brief overview:

1. We add [django-celery-beat] (an official celery project) as an additional dependency to GeoNode. It is able to store celery beat schedule in the django database and exposes models to update this schedule in a dynamic way. Each scheduled task is represented by a `PeriodicTask` model and it has properties such as the update frequency, whether it is enabled or not, which celery task shall be executed, etc.

2. Then I have a `Harvester` model, which is connected to `PeriodicTask` via a 1-1 relationship - this means that every harvester has an associated periodic task, and is therefore able to be scheduled by celery beat in a dynamic way. Each harvester keeps track of its own properties

This seemed to work really well when I did the initial development. However, when I started adding more harvesters I realized that the scheduling does not seem to be working correctly. I seem to be facing an issue similar to what is described here:

https://github.com/celery/django-celery-beat/issues/330

Basically, I am seeing that only two periodic tasks seem to be able to run at the same time. This seems like a strange bug. I plan to debug it in order to see if it can be made to work reliably. As such, it is currently unclear to me how much time/effort this will still require. I will know better when I have the time to do some debugging.

[django-celery-beat]: https://github.com/celery/django-celery-beat

**Comment by ricardogsilva** [comment]
@meomancer 

can you please debug this to see if it is still relevant?

There have been new releases of `django-celery-beat`, so maybe this has been solved in the meanwhile?

For debugging you will need to:

- Make sure you run a celery beat scheduler by launching celery with `celery --app geonode.celery_app beat -l DEBUG ` as the scheduler uses celery beat
- Add some three or four harvester instances to the django DB, 
- Set their `scheduling_enabled` attribute to `True`
- Set their `update_frequency` attribute to a value suitable for testing - this is interpreted as minutes
- Monitor the celery beat logs in order to check whether all harvesters are being scheduled. When I originally implemented this code I found that some of the harvesters were triggered but others were not

You can try to update the django-celery-beat version mentioned in the geonode `requirements.txt` and `setup.cfg` to the latest version and check whether that fixes the issue

**Comment by meomancer** [comment]
Hi @ricardogsilva 
It is still just running the first 2 task from harvesters (using latest celery and celerybeat)
But i'm still not be able to debug it and still can't find the problem

**Comment by giohappy** [comment]
can we close this @ricardogsilva?

**Comment by ricardogsilva** [comment]
@giohappy , yes a dynamic scheduler has been implemented and merged upstream. I'll go ahead and close this
